### ISSUE TITLE
DriveTransformer: Unified Transformer for Scalable End-to-End Autonomous Driving

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#66]
<img width="1406" alt="Image" src="https://github.com/user-attachments/assets/1583d291-7de6-46da-8ac2-d109ec6e57a6" />

### Paper Link
https://openreview.net/forum?id=M42KR4W9P5

## Challenges in E2E AD

<img width="1344" alt="Image" src="https://github.com/user-attachments/assets/7f1bb5d3-5b9d-4506-bef8-ecea3b071afc" />

**Sequential pipeline**
- Existing E2E-AD methods mostly adopt a sequential pipeline of perception-prediction-planning, where **downstream tasks are heavily dependent on upstream queries**. This sequential design can lead to **cumulative errors** and thus training instability. 
- For instance, the training process of UniAD necessitates a **multi-stage** approach: first, pre-training the BEVFormer encoder; then, training TrackFormer and MapFormer; and finally, training MotionFormer and Planner. This fragmented training approach **increases the complexity and difficulty of deploying and scaling the system** in industrial settings. 
- Moreover, the manual ordering of tasks can restrict the system’s ability to leverage synergies, such as planning-aware perception.

**Parallel pipeline**
- Latest work ParaDrive (CVPR2024) tries to mitigate the instability issue by **removing all tasks’ connection**. However, it still suffers from the **expensive BEV representation**.

### DriveTransformer in this paper
- Task Parallelism: **All task queries directly interact with each other at each block**, fostering **crosstask knowledge transfer** while maintaining system stability without explicit hierarchy.
- Sparse Representation: **Task queries directly engage with raw sensor features**, offering an efficient and direct means of information extraction, aligning with end-to-end optimization paradigm.
- Streaming Processing: **Temporal fusion** is achieved through a first-in-first-out queue that stores task queries in history and temporal cross attention, ensuring efficiency and feature reuse.

<img width="1500" alt="Image" src="https://github.com/user-attachments/assets/3edc7c6f-fbb4-4520-8e70-c0c4079030df" />

### Initialization & Tokenization

<img width="1610" alt="Image" src="https://github.com/user-attachments/assets/c0c287df-764a-4cff-9f4c-8480fabea157" />

**Image Tokens:**

- Prior to information exchange in DriveTransformer, all inputs are converted into a unified representation - tokens.
- All tokens consist of two parts: **semantic** embeddings for semantic information and **position** encodings for spatial localization.

**Task Tokens:**
1. **Agent Queries** represent dynamic objects (vehicles, pedestrians, etc), which will be used to conduct object detection and motion prediction.
2. **Map Queries** represent static elements (lanes, traffic signs, etc), which will be used to conduct online mapping.
3. **Ego Query** represents the potential behavior of the ego vehicle, which will be used to conduct planning.

> **Technique Details**: Backbone is ResNet. 3D Position Encoding is adopted to encode the position information of sensor features. Coordinates of points in the same ray are concatenated and fed into an MLP to obtain the position encoding for each patch

### Initialization & Tokenization

- **Sensor Cross Attention (SCA)** establishes a direct pathway between tasks and raw sensor inputs, enabling end-to-end learning without information loss.
- **Task Self-Attention (TSA)** enables direct interaction among arbitrary tasks without explicit constraint, promoting synergy such as planning-aware perception and gametheoretic interactive prediction and planning.
- **Temporal Cross Attention** integrates information from previously observed history.

<img width="1390" alt="Image" src="https://github.com/user-attachments/assets/e8bb3d0b-dbe2-4919-812a-299fc1e29fb0" />

**Temporal Cross Attention**
- Motivation: a) Maintaining long-term BEV features is expensive. b) Previous task queries carrying strong prior semantic and spatial information are wastefully discarded.

> K: 50, T: 10 for Bench2Drive (1 second at 10Hz) and 4 for nuScenes (2 seconds at 2Hz)

### Task Head
<img width="1143" alt="Image" src="https://github.com/user-attachments/assets/7e85a4db-e14c-4ad1-9e22-8cad329d3def" />
- **Object Detection & Motion Prediction**: conducting object detection and motion prediction without tracking, by feeding the same agent query into different task heads. The same feature for the same agent naturally establish associations between detection and prediction.
-** Online Mapping**: We replicate each map query for N times paired with position encodings for each single point. For those long polylines, each point could retrieve raw sensor information with better locality. To integrate separate point-level map queries into instance-level ones for other modules, we adopt a light-weight PointNet (Qi et al., 2017) with max-pooling and MLPs.
- **Planning**: We model the the future movement of the ego vehicle as Gaussian Mixture Model to avoid mode averaging. **Training**: Go Straight, Stop, Left Turn, Sharp Left Turn, Right Turn, Sharp Right Turn -> predict six mode-specific ego trajectories. -> only the Ground-Truth mode’s trajectory would be used to calculate regression loss; **Inference**: the trajectory from the mode with the highest confidence score would be used to compute metrics or execute.
- **Coarse-to-Fine Optimization**: the Map and Agent PE are encoded by their corresponding **predicted positions** and semantic classes with an MLP to capture the spatial and semantic relations among elements. The ego PE is encoded by the predicted planning trajectory with an MLP to capture the ego intentions for the possible interactions.

## Experiment
<img width="1311" alt="Image" src="https://github.com/user-attachments/assets/45321ff6-0cd0-4d63-b04a-ba30a596d5ae" />

<img width="694" alt="Image" src="https://github.com/user-attachments/assets/0d70109c-27b1-4c24-9e8d-877e67aa2894" />

<img width="1027" alt="Image" src="https://github.com/user-attachments/assets/ab23b80e-e87e-4644-9be3-77a74374daee" />

<img width="1131" alt="Image" src="https://github.com/user-attachments/assets/f5395606-4eea-4ed6-8b82-cccbd8610aef" />

**Closed-Loop Evaluation** involves deploying the model in a simulated or real environment where its actions (e.g., steering, braking) affect subsequent inputs, allowing you to assess its long-term driving performance and safety. 

**Open-Loop Evaluation** is an offline analysis where the model's predictions are compared against ground truth trajectories without influencing the environment, focusing on metrics like trajectory error and collision rates without real-time feedback.
